cmd/snap/cmd_info.go

```diff
@@ -163,27 +164,78 @@ func coalesce(snaps ...*client.Snap) *client.Snap {
 	return nil
 }
 
-// formatDescr formats a given string (typically a snap description)
+// runesTrimRightSpace returns text, with any trailing whitespace dropped.
+func runesTrimRightSpace(text []rune) []rune {
+	j := len(text)
+	for j > 0 && unicode.IsSpace(text[j-1]) {
+		j--
+	}
+	return text[:j]
+}
+
+// runesLastIndexSpace returns the index of the last whitespace rune
+// in the text. If the text has no whitespace, returns -1.
+func runesLastIndexSpace(text []rune) int {
+	for i := len(text) - 1; i >= 0; i-- {
+		if unicode.IsSpace(text[i]) {
+			return i
+		}
+	}
+	return -1
+}
+
+// wrapLine wraps a line to fit into width, preserving the line's indent, and
+// writes it out prepending padding to each line.
+func wrapLine(out io.Writer, text []rune, pad string, width int) {
+	// Note: this is _wrong_ for much of unicode (because the width of a
+	//       rune on the terminal is anything between 0 and 2, not always 1
+	//       as this code assumes) but fixing that is Hard. Long story
+	//       short, you can get close using a couple of big unicode tables
+	//       (which is what wcwidth does). Getting it 100% requires a
+	//       terminfo-alike of unicode behaviour.
+	// This (and possibly printDescr below) should move to strutil once
+	// we're happy with it getting wider (heh heh) use.
+
+	// discard any trailing whitespace
+	text = runesTrimRightSpace(text)
+	// establish the indent of the whole block
+	idx := 0
+	for idx < len(text) && unicode.IsSpace(text[idx]) {
+		idx++
+	}
+	indent := pad + string(text[:idx])
+	text = text[idx:]
+	width -= idx + utf8.RuneCountInString(pad)
+	for len(text) > width {
+		// find a good place to chop the text
+		idx = runesLastIndexSpace(text[:width+1])
+		if idx < 0 {
+			// there's no whitespace; just chop at line width
+			idx = width
+		}
+		fmt.Fprint(out, indent, string(text[:idx]), "\n")
+		// prune any remaining whitespace before the start of the next line
+		for idx < len(text) && unicode.IsSpace(text[idx]) {
+			idx++
+		}
+		text = text[idx:]
+	}
+
+	fmt.Fprint(out, indent, string(text), "\n")
+}
+
+// printDescr formats a given string (typically a snap description)
 // in a user friendly way.
 //
 // The rules are (intentionally) very simple:
-// - trim whitespace
-// - word wrap at "max" chars
-// - keep \n intact and break here
-// - ignore \r
-func formatDescr(descr string, max int) string {
-	out := bytes.NewBuffer(nil)
-	for _, line := range strings.Split(strings.TrimSpace(descr), "\n") {
-		if len(line) > max {
-			for _, chunk := range strutil.WordWrap(line, max) {
-				fmt.Fprintf(out, "  %s\n", chunk)
-			}
-		} else {
-			fmt.Fprintf(out, "  %s\n", line)
-		}
+// - trim trailing whitespace
+// - word wrap at "max" chars preserving line indent
+// - keep \n intact and break there
+func printDescr(w io.Writer, descr string, max int) {
```

Comment:
Similar comment about io.Writer. As a general hint: if it takes an io.Writer, it needs to return an error.

Reply:
I understand. However this is just two of eleven similar functions `cmd/snap` (9 of them in `cmd_info.go` itself); if we want to change we should do it in a separate PR.